### DISCUSSION TITLE
Feasability of changing hmmpr to tmmpr for Inch compatibility

### DISCUSSION THREAD

**Discussion by gstwrt** [kachurovskiy/nanoels#75] [Ideas]
I've been looking through the code with thoughts of adding an inch/tpi mode.   To me the most sensible way to do this would be to change the core value of hundredths of a mm per rev to thousands, then repurpose a shortcut button or two for setting in in/tpi.  Everything in the code would stay in mm, but the button would toggle a flag/multiplier for inch measurements.   The reason for the change to thousandths is that 25.4 mm per inch requires 1 more significant figure of precision to cover the common thread pitches.

I have enough experience to make the code changes myself, though I'm new to github and contributing to a shared code project. 

My questions are if there are any likely overflow or performance problems with changing the core unit from hundredths to thousandths of a mm?

**Comment by kachurovskiy** [comment]
I vaguely recall that there's an enormous amount of different imperial standards, each having a slightly different definition for e.g. 1/2" pitch. E.g. BSB vs BSF vs ... - https://www.gewinde-normen.de/en/bsb-thread.html - how do we handle all that complexity?

> The reason for the change to thousandths is that 25.4 mm per inch requires 1 more significant figure of precision to cover the common thread pitches

I'd like to learn more about this. Why is having 3 precision digits required? With current logic, you can get your pitch within 0.005mm of desired. That gives ~1% error on some imperial pitches:

![image](https://user-images.githubusercontent.com/517919/195427179-f6b6797e-781c-434f-bf70-7bfd122fd24e.png)

That does look concerning but I wonder what are the official tolerances for thread pitches? For metric is appears to be around 1%.

How will the third precision digit be handled in metric mode? Currently a click on + or - button adds/substracts 0.01mm from pitch.  Keeping it at 0.01mm - how will the third digit be controlled in the metric mode? Changing it to add/substract 0.001mm would require too much effort to change pitches. Maybe we could add/substract 0.001mm when the `D` button is pressed while pressing + or -.

How will cycling through the imperial pitches be done?

What would be the actual list of TPI and Inch threads supported?

What happens when we switch from metric to imperial mode and `tmmpr` doesn't match any imperial thread?

We have `S` button controlling step - 1mm, 0.1mm and 0.01mm - how will it be affected by the imperial mode?

> My questions are if there are any likely overflow or performance problems with changing the core unit from hundredths to thousandths of a mm?

This should not be a problem IIUC.

**Reply by gstwrt** [reply]
Wow.  Thanks for the detailed response.

Yes, imperial thread standards are a slight mess, though mostly they all use integer tpi.   Differences are in major and minor diameters and angles - which are set by the tool and compound angle.   

I didn't realize the error was generally so low for most thread sizes using hundredths of a mm.   I'd checked 12tpi and decided the error was too much. Changing to thousandths of a mm internally also reduced cumulative errors on position when in inch mode or manually setting the pitch.

I spent a bit of time modifying the code today.   Here is how I handled all those issues:

I implemented three modes.  mm, in and tpi, with the last being just a different way of showing pitch.   Switching modes (B5) changes the movesteps, pitch list, display units, and decimal precision.  B3 cycles through a list of pitches in the appropriate units.  When using the plus minus buttons in inch mode, the pitch increment and decrement size is changed to 0.001in. I'm working on getting plus&minus to work in tpi mode, but I've got some issues around 0 crossing.   B2 still cycles though movestep sizes, though I've changed the way it is implemented.

When switching modes,  tmmpr doesn't change - allowing the use of mixed units.   Ie.  it is possible to set a thread pitch in inches or tpi and then display the values in mm while cutting.

Instead of Defs for move step and pitch, I've implemented them in a const multidimensional array.  It can be arbitrarily long.   Uses a tiny bit more ram.  

As I mentioned I'm new to github so I'm not super familiar with how to get a copy of the changes back upstream.  I'll see if I can figure out a pull request, but please excuse me if I do it wrong at first.

**Reply by kachurovskiy** [reply]
@gstwrt thank you for sharing your work.

This seems like a significant departure from what we currently have in H2 and also potentially a more complex UI to use. I'd love to give it a try and take a look at the code but overall I would prefer this to be a separate codebase e.g. in your GitHub repo which I'll link to for imperial fans.

GitHub-wise, all you need to do is to fork this repository into your account and push commits into it (no pull request required for now). Thank you!

**Comment by gstwrt** [comment]
The UI really isn't more complex.   When working in mm mode, the only UI change is that there is 1 button to choose a preset pitch instead of 3 shortcuts.  B5 switches units and everything else is functionally the same.

I'd love for the code to be integrated into the main project, but your project, your call of course.    I'll fork and in the next hour or two and you can try it out.

**Reply by kachurovskiy** [reply]
It's the beauty and power of open source that we can have more than one thing to chose from. I think your version of NanoEls would be much more interesting and would develop faster if you don't have to ask my permission for every change.

**Comment by gstwrt** [comment]
![20221012_181014](https://user-images.githubusercontent.com/115626897/195464603-75d78ec0-1a6c-40dd-9a66-529cb35b2ff4.jpg)
https://github.com/gstwrt/nanoels-imperial

h2.ino is the only changed file.  Still a few bugs and I haven't yet dealt with the eeprom.

**Reply by kachurovskiy** [reply]
Looks nice! Keep us updated.

**Reply by kachurovskiy** [reply]
Please let me know when it's OK to advertise your version on the front page :)

**Comment by gstwrt** [comment]
Absolutely.    There are still a few bugs to sort out, which I will hopefully be able to take care of this weekend.

**Comment by gstwrt** [comment]
Here is something you could maybe help me with.   I thought I had somehow introduced a bug which was causing the position to move by a larger than expected amount when the moveStep is smaller than the pitch and stepper is on.   I'm looking at the logic in checkMoveButtons() and I've started thinking that it is working how you designed it, but I can't wrap my head around why or what the exact logic is in the lines:       
```
// Move by moveStep in the desired direction but stay in the thread by possibly traveling a little more.
      spindlePos += ceil(MOTOR_STEPS * moveStep * 1.0 / LEAD_SCREW_HMM * STEPPER_TO_ENCODER_STEP_RATIO / ENCODER_STEPS / abs(hmmpr))
                    * ENCODER_STEPS
                    * sign
                    * (hmmpr > 0 ? 1 : -1);
```

**Reply by kachurovskiy** [reply]
The goal here is to stay "in" the thread - to do so, I'm pretending to rotate the spindle by N whole rotations (.., -2, -1, 1, 2, ...). Once we "rotate" the spindle, the default stepper moving logic does the rest.

The exact math of finding the number of whole rotations (e.g. 2) is such that we try to:
- Move the axis at least 1mm/0.1mm/0.01mm (based to the "Step" set)
- Keep the number of imaginary spindle rotations a round number that's not 0 (hence `ceil()`)

Looking at these lines I have hard time fully parsing them myself honestly :) It just takes time, I've probably put 100 hours in that 1000 lines of code.

**Comment by gstwrt** [comment]
I think I have all the important bugs worked out, though my changes lack on lathe testing.   My 3d printer is not currently working so I haven't yet installed the hardware on my lathe.

I'd love if some one could test it IRL and provide feedback.

I also realized earlier that it would be quite simple to add a single define block in secondarywork() and create a metric only mode with the original function buttons restored.

**Reply by kachurovskiy** [reply]
I can certainly help testing the final version but I'm afraid I don't have the energy to do more than a few iterations. I've looked though your published commits and don't see any obvious problems. Great to hear about your progress 👍

**Comment by gstwrt** [comment]
Thanks.   I didn't make any significant changes to the step generation sections of the code so it should work as before.   Mostly, it would be great if someone could check and make sure I didn't somehow introduce a factor of 10 error when changing from hundredths to thousandths of a mm.   If you are too busy (I completely understand that) I can make a separate discussion post asking for testers.   I'm hoping to have the encoder and stepper set up on my lathe in the next couple weeks if life doesn't interfere.

**Reply by kachurovskiy** [reply]
I've actually implemented thousands in the main repository as well by now and added a bunch more improvements - https://github.com/gstwrt/nanoels-imperial/compare/main...kachurovskiy:nanoels:main - thank you for motivating me to get things done. Hope you'd be able to merge them as well to make the inch UI version even better.

Testing stuff is core to this work and I have to re-test every change dozens of times before it works, I'm not sure this can be outsourced. Feel free to ask around though.
